README.md
![Version](https://img.shields.io/badge/Version-1.2-brightgreen.svg)
![GhostVersion](https://img.shields.io/badge/GhostVersion-0.11.7-red.svg)

# ghostPostRetriever
A Ghost blog's posts retrieve engine. Shows posts with pagination. Optionally pages are switched without reloading and with keeping the browser history.


**Original developer:** [skyweb.piotr.gabara@gmail.com](mailto:skyweb.piotr.gabara@gmail.com)

ghostPostRetriever is a Javascript engine for retrieving posts using [Ghost API](https://api.ghost.org/v0.1/docs).

## Basic setup

```txt
git clone https://github.com/petergab/ghostPostRetriever.git --recursive
```

After cloning, the ghostPostRetriever module will be located at `ghostPostRetriever/jquery.ghostPostRetriever.js`. Remember to minify it for production use.

To test the module in your template, add the following line, after JQuery is loaded. Typically this will be near the bottom of a file `default.hbs`, in the top folder of the theme directory.

```html
<script type="text/javascript" src="{{asset "js/ghostPostRetriever/jquery.ghostPostRetriever.js"}}"></script>
```

You will need to add a container for filtered posts with the specified id to your pages.

```html
<section id="posts-container"></section>
```

Implement ghostPostRetriever with a block of JQuery code.

```html
<script type="text/javascript">
  $("#posts-container").ghostPostRetriever();
</script>
```


### GhostPostRetriever options

GhostPostRetriever offers a set of options when the plugin is invoked.

:arrow_right: **postTemplate**
> A simple Handlebars template used to render retrieved posts. The template uses variable substitution only. Helpers and conditional insertion constructs are ignored and will be rendered verbatim.
>
> Default template is just simple:
```html
 <a href="{{link}}">{{title}}</a>
```

:arrow_right: **tagTemplate**
> A Handlebars template used to display Tags (if included in getAdditionalPostOptions).
>
> Default usage: Just put ``{{tagList}}`` somewhere in your template
>
> Default template is:
```html
<a href="/tag/{{slug}}">{{name}}</a>
```

:arrow_right: **authorTemplate**
> A Handlebars template used to display Authors (if included in getAdditionalPostOptions).
>
> Default usage: Just put ``{{authorViaTemplate}}`` somewhere in your template
>
> Default template is:
```html
<a href="/tag/{{slug}}">{{name}}</a>
```

:arrow_right: **paginationContainerTemplate**
> A Handlebars template used to display pagination.
>
> Default template is:
```html
<div class="extra-pagination inner">
  <nav class="pagination">
    {{prevTemplate}}
    <span>Page {{page}} of {{pages}}</span>
    {{nextTemplate}}
  </nav>
</div>
```

:arrow_right: **paginationPrevTemplate**
> A Handlebars template used to display link for previous page in pagination template.
>
> Default template is:
```html
<a href="{{urlPrev}}">&larr; Newer Posts</a>
```

:arrow_right: **paginationNextTemplate**
> A Handlebars template used to display link for next page in pagination template.
>
> Default template is:
```html
<a href="{{urlNext}}">Older Posts &rarr;</a>
```

:arrow_right: **paginationContainer**
> JQuery ID of the DOM object into which pagination container should be inserted. If not set the pagination container will be inserted at the end of the posts results.
>
> Default value is: ``false``

:arrow_right: **paginationShow**
> The flag to decide if the pagination should be visible
>
> Default value is: ``true``

:arrow_right: **before**
> Use to optionally set a callback function that is executed before the list of results is displayed. The callback function takes no arguments.
>
> Default value is: ``false``
>
> Example:
```javascript
$("#posts-container").ghostPostRetriever({
  before: function() {
    alert("Posts are about to be rendered");
  }
});
```

:arrow_right: **onComplete**
> Use to optionally set a callback function that is executed after the list of results is displayed. The callback function takes no arguments.
>
> Default value is: ``false``

:arrow_right: **page**
> The page number that you'd like to display
>
> Default value is: ``1``

:arrow_right: **postsLimit**
> The posts limit on single page
>
> Default value is: ``6``

:arrow_right: **toLocaleDateStringLocale**
> A string with a BCP 47 language tag, or an array of such strings used in [The toLocaleDateString()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString) method to show publication date.
>
> Default value is: ``'en-US'``
>
> Default usage: Just put ``{{prettyPubDate}}`` somewhere in your template instead of ``published_at``

:arrow_right: **toLocaleDateStringOptions**
> An object with additional properties used in [The toLocaleDateString()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString) method.
>
> Default value is:
```javascript
{
  year: 'numeric',
  month: 'long',
  day: '2-digit'
}
```

:arrow_right: **getAdditionalPostOptions**
> You can make your API queries more selective using a range of parameters described in the [TGhost API documentation](https://api.ghost.org/docs/parameters).
>
> Default value is: ``{}``
>
> Example
```javascript
{
  include: 'author, tags',
  filter: 'tags:[your-tag]'
}
```

:arrow_right: **exerptLimit**
> If you'd like to present only sneak peak of your post you can set the number of letters to take from the html value
>
> Default value is: ``200``
>
> Default usage: Just put ``{{exerpt}}`` somewhere in your template

:arrow_right: **zeroResultsInfo**
> Information to be shown when there are no posts to be shown (also the pagination container is not presented)
>
> Default value is: ``'There are no posts that meet specified criteria.'``

:arrow_right: **clearTargetBeforeInserting**
> The flag to decide if you want to clear the target before inserting results
>
> Default value is: ``true``


changelog.txt
v1.2 : More ES6
v1.1.3 : Added flag to decide if you want to clear the target before inserting results
v1.1.2 : FIX: Remove unnecessary parameter
v1.1.1 : FIX: ability to choose only few fields from each post
v1.1 : Added flag to decide if the pagination should be visible
v1.0 : Added ability to switch pages without reloading the page and with keeping the the browser history
v0.1.2 : ESLint fixes
v0.1.1 : Added missing parameters: onComplete, zeroResults, paginationContainer
v0.1 : Initial commit, first release

jquery.ghostPostRetriever.js
/*global jQuery, ghost*/
/*eslint max-statements: ["error", 25]*/
/*eslint no-use-before-define: ["error", { "variables": false }]*/

/**
* ghostPostRetriever - 1.2
 * for Ghost Version: 0.11.7
 * Copyright (C) 2019 Piotr Gabara (skyweb.piotr.gabara@gmail.com)
 * https://github.com/petergab/ghostPostRetriever
 * MIT Licensed
 * @license
*/
(function($) {

  $.fn.ghostPostRetriever = function(options) {
    const opts = $.extend({}, $.fn.ghostPostRetriever.defaults, options);
    pluginMethods.init(this, opts);
    return pluginMethods;
  };

  $.fn.ghostPostRetriever.defaults = {
    postTemplate: '<a href="{{link}}">{{title}}</a>',
    tagTemplate: '<a href="/tag/{{slug}}">{{name}}</a>',
    authorTemplate: '<a href="/author/{{slug}}">{{name}}</a>',
    paginationContainerTemplate: '' +
      '<div class="extra-pagination inner">' +
        '<nav class="pagination">' +
          '{{prevTemplate}}' +
          '<span> Page {{page}} of {{pages}}</span> ' +
          '{{nextTemplate}}' +
        '</nav>' +
      '</div>',
    paginationPrevTemplate: '<a id="pagination-prev" href="{{urlPrev}}">&larr; Newer Posts</a>',
    paginationNextTemplate: '<a id="pagination-next" href="{{urlNext}}">Older Posts &rarr;</a>',
    paginationContainer: false,
    paginationShow: true,
    before: false,
    onComplete: false,
    page: 1,
    postsLimit: 6,
    toLocaleDateStringLocale: 'en-US',
    toLocaleDateStringOptions: {
      year: 'numeric',
      month: 'long',
      day: '2-digit'
    },
    getAdditionalPostOptions: {},
    exerptLimit: 200,
    zeroResultsInfo: 'There are no posts that meet specified criteria.',
    clearTargetBeforeInserting: true
  };

  const pluginMethods = {
    init: function(target, opts) {
      this.target = target;
      this.postTemplate = opts.postTemplate;
      this.tagTemplate = opts.tagTemplate;
      this.authorTemplate = opts.authorTemplate;
      this.paginationContainerTemplate = opts.paginationContainerTemplate;
      this.paginationPrevTemplate = opts.paginationPrevTemplate;
      this.paginationNextTemplate = opts.paginationNextTemplate;
      this.paginationContainer = opts.paginationContainer;
      this.paginationShow = opts.paginationShow;
      this.before = opts.before;
      this.onComplete = opts.onComplete;
      this.page = opts.page || 1;
      this.postsLimit = opts.postsLimit;
      this.toLocaleDateStringLocale = opts.toLocaleDateStringLocale;
      this.toLocaleDateStringOptions = opts.toLocaleDateStringOptions;
      this.exerptLimit = opts.exerptLimit;
      this.zeroResultsInfo = opts.zeroResultsInfo;
      this.clearTargetBeforeInserting = opts.clearTargetBeforeInserting;

      const getPostOptionsDefault = {
        limit: this.postsLimit,
        page: parseInt(this.page) || 1
      };
      this.getPostOptions = $.extend({}, getPostOptionsDefault, opts.getAdditionalPostOptions);
      this.setObservers();

      if (this.before) {
        this.before();
      }

      this.getAndShowPosts();
    },

    setObservers: function() {
      if (this.paginationShow) {
        let paginationWrapper;
        if (this.paginationContainer && this.paginationContainer.length > 0) {
          paginationWrapper = this.paginationContainer;
        } else {
          paginationWrapper = this.target;
        }
        // The observer is set on the paginationContainer if it exists, if not it is set on whole module target container
        $(paginationWrapper).on('click', '#pagination-prev, #pagination-next', (e) => this.paginationClick(e));
      }

      window.onpopstate = (e) => {
        this.getPostOptions.page = ((e.state && e.state.page) ? e.state.page : 1);
        this.getAndShowPosts();
      };
    },

    paginationClick: function(e) {
      e.preventDefault();
      if (($(e.target).attr('id') === 'pagination-prev') && (parseInt(this.page) > 1)) {
        this.page = parseInt(this.page) - 1;
      }
      if (($(e.target).attr('id') === 'pagination-next') && (parseInt(this.page) > 0)) {
        this.page = parseInt(this.page) + 1;
      }
      this.getPostOptions.page = this.page;
      this.getAndShowPosts();
      window.history.pushState({page: this.page}, '', `${window.location.pathname}?page=${this.page}`);
      return false;
    },

    getAndShowPosts: function() {
      if (this.clearTargetBeforeInserting) {
        this.clear();
      }
      $.get(ghost.url.api('posts', this.getPostOptions)).done((data) => {
        let resultHtml = '';

        data.posts.forEach((post) => {
          if (post.html) {
            post.exerpt = this.createExcerpt(post.html);
          }
          if (post.published_at) {
            post.prettyPubDate = this.prettyDate(post.published_at);
          }
          if (post.author) {
            post.authorViaTemplate = this.format(this.authorTemplate, post.author);
          }
          if (post.tags) {
            const tagsArray = post.tags.map((t) => {
              return (t.visibility === 'public' ? this.format(this.tagTemplate, t) : '');
            });
            post.tagList = (tagsArray.length > 0 ? tagsArray.filter(Boolean).join(', ') : '');
          }
          resultHtml += this.format(this.postTemplate, post);
        });

        const pagination = data.meta.pagination;
        if (pagination.total > 0) {
          if (this.paginationShow) {
            pagination.urlPrev = (pagination.prev ? window.location.pathname + '?page=' + pagination.prev : '');
            pagination.urlNext = (pagination.next ? window.location.pathname + '?page=' + pagination.next : '');
            pagination.prevTemplate = (pagination.urlPrev ? this.format(this.paginationPrevTemplate, pagination) : '');
            pagination.nextTemplate = (pagination.urlNext ? this.format(this.paginationNextTemplate, pagination) : '');

            if (this.paginationContainer) {
              $(this.paginationContainer).html(this.format(this.paginationContainerTemplate, pagination));
            } else {
              resultHtml += this.format(this.paginationContainerTemplate, pagination);
            }
          }
        } else {
          resultHtml = this.zeroResultsInfo;
        }

        $(this.target).append(resultHtml);

        if (this.onComplete) {
          this.onComplete();
        }
      }).fail((err) => {
        $(this.target).append(`Error ${err} occured. Please try again.`);
      });
    },

    clear: function() {
      // Setting the container min-height so it doesn't jump on switching content
      $(this.target).css('min-height', `${$(this.target).height()}px`);
      $(this.target).empty();
    },

    prettyDate: function(date) {
      const d = new Date(date);
      return d.toLocaleDateString(this.toLocaleDateStringLocale, this.toLocaleDateStringOptions);
    },

    createExcerpt: function(text) {
      return text.replace(/<\/?[^>]+(>|$)/g, '').replace(/\n/, ' ').substring(0, this.exerptLimit) + '...';
    },

    format: function(template, data) {
      return template.replace(/{{([^{}]*)}}/g, function(a, b) {
        const r = data[b];
        return typeof r === 'string' || typeof r === 'number' ? r : a;
      });
    }
  };

})( jQuery );
